messaging/WFLY-13793_disable_legacy_amq1_prefix.adoc

```diff
@@ -0,0 +1,128 @@
+= [WFLY-13793] Allow for a remote jms queue / topic not to use legacy amq1 prefix
+:author:            Chao Wang
+:email:             chaowan@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:          messaging,jms
+:idseparator:       -
+
+== Overview
+Currently we can disable the use of legacy amq1 prefixes `jms.queue` and `jms.topic` only at the connection factory level but not for queue or topic defined in the model or using JakartaEE annotations.
+The goal of this proposal is to be able to disable the usage of such prefixes at the destination level for external JMS queues and topics.
```

Comment:
@soul2zimate I realized that also the final behavior is not clear to me and that we could probably add some info here; 

Let me use an example: let's say we have a queue defined in the model like :

```
<external-jms-queue name=myExternalQueue" entries="java:jboss/exported/jms/queue/myExternalQueue" enable-amq1-prefix="true"/>
```

and that the PCF is configured with the opposite setting:

```
/subsystem=messaging-activemq/pooled-connection-factory=remote-artemis:write-attribute(name="enable-amq1-prefix", value="false")
```

we said that PFC has no effect when creating the queue hence I guess the queue gets created with name `jms.queue.myExternalQueue`;

what happens when sending a message using the PFC? will it try to remove the `jms.queue.` prefix?

is the following configuration equivalent? is it even correct?: 

```
<external-jms-queue name=jms.quque.myExternalQueue" entries="java:jboss/exported/jms/queue/myExternalQueue" enable-amq1-prefix="true"/>
```

Reply:
The PCF doesn't remove prefixes, it just don't append them

Reply:
> is the following configuration equivalent? is it even correct?:
>
><external-jms-queue name=jms.quque.myExternalQueue" entries="java:jboss/exported/jms/queue/myExternalQueue" enable-amq1-prefix="true"/>

The remote queue will be consisted of the prefix `jms.queue` + queue name `jms.quque.myExternalQueue`, so   it will create a remote queue with name `jms.queue.jms.queue.myExternalQueue`. But I don't think this is best practice to do so (having two `jms.queue` in the front)

Reply:
I was just asking because it wasn't clear to me how this `enable-amq1-prefix` works: IIUC when set to true it always adds the prefix even if it's already present, when set to false it doesn't remove the prefix even if already preset, correct?

Reply:
@tommaso-borgato you are talking about different things. In WildFly we are only manipulating the address used to be passed to Artemis RA or API: so it out of the scope of this RFE how this RA works. The 'only' place where we really do something significant is at the queue creation command which is but a subset of the current RFE. This feature is mostly about beeing able to call and use a queue called **myQueue** on a remote Artemis without having to use an external jndi context.

Reply:
@ehsavoie I try and summarize the current status: feature has "accidentally" been merged without the AD being approved nor the pre-check being granted....
Now we can either: revert that change or try and complete the works here; IIRC, we stopped because it was unclear how the whole thing worked and we could not picture it out within the time we had available back then;

IIRC, we were talking about the different `points` where the queue/topic name can be manipulated (= the legacy amq1 prefix can be added/removed):
- inside the deployed application JAR/WAR/EAR: e.g. Jakarta EE annotation (can we also use deployment descriptor? other?)
- inside the model when defining the item: e.g. external-jms-queue
- inside the model when defining the how to communicate with the internal/external broker: e.g. pooled-connection-factory

is that all? could you start by refining and/or completing this list?

I would then move on with the remaining questions:
- what's the priority for the `points` mentioned above?
- where can the legacy amq1 prefix be added/removed? can it just be added?
- what difference does it make if the broker is external/internal?
- what difference does it make if the queue/topic already exists or not?
- what happens if the queue/topic doesn't exist? is it auto-created? is the prefix added?

Reply:
@ehsavoie just copying an pasting your answer here to make the thread consistent:

this concerns only 'external" queues and topics as the internal EAP broker enforces the use of the legacy prefixes for locally defined destinations.
We need to enable this at the model and at the deployment level.

You have different ways to access an external broker that's why we need the attribute at different level. It all depends on how you are obtaining the destination:

    locally defined and injected: here you need it at the communication level and at the destination level
    remote JNDI call on the broker: here you need it at the communication level
    usage of Artemis API and destination address: here you need it at the communication level

The legacy prefix is here to ensure retro-compatibility with the destination addresses defined in HornetQ and Artemis 1.x. See [Fully Qualified Queue Names](https://activemq.apache.org/components/artemis/documentation/latest/address-model.html) to understand the impact of those prefixes.
Setting the attribute at the communication level isn't sufficient as WildFly is enforcing the use of the legacy prefixes on the destination address.
We can configure the broker to 'ignore' the prefixes automatically but this won't be applied at the creation of the destinations by WildFly. So while we would be using the JMS queue myqueue we would still create a jms.queue.myqueue*.
A queue can be autocreated by the broker depending on the broker configuration, by WildFly can do it also if the broker won't allow it. We need to ensure that the address that WildFly is using matches that of the broker. The setting of the prefix is not only done by Wildfly but also by the Artemis client and resource adapter.